### PR TITLE
[master < E219] Map projection

### PR DESCRIPTION
**Pull request by antepusic**
[master < Epic] PR
- [x] Check, and update documentation if necessary
- [x] Write E2E tests
- [x] Compare the [benchmarking results](https://bench-graph.memgraph.com/) between the master branch and the Epic branch
- [x] Provide the full content or a guide for the final git message
> Add the map projection feature

* NB: Benchmarks compared between master and #931 because that branch builds on all other task branches and thus has all the new code

To keep docs changelog up to date, one more thing to do:
- [x] Write a release note here
- [x] Tag someone from docs team in the comments

Release note:
> It’s now possible to use the map projection syntax to create maps. Map projections are convenient for building maps based on existing values and they are used by data access tools like GraphQL and Spring Data.

### CODE REVIEW

File: src/query/frontend/ast/ast.hpp [hunk 1]
```diff
@@ -1087,6 +1088,60 @@ class MapLiteral : public memgraph::query::BaseLiteral {
...
+        if (!pair.second) continue;
+
+        if (!pair.second->Accept(visitor)) break;
+      }
+    }
+    return visitor.PostVisit(*this);
+  }
+
+  Expression *map_variable_;
+  std::unordered_map<PropertyIx, Expression *> elements_;
```
Comment (Josipmrden) [inline src/query/frontend/ast/ast.hpp:1118]: can't we wrap this as MapProjectionData?

File: src/query/frontend/ast/pretty_print.cpp [hunk 2]
```diff
@@ -122,6 +126,15 @@ void PrintObject(std::ostream *out, Expression *expr) {
   }
 }
 
+void PrintObject(std::ostream *out, AllPropertiesLookup *apl) {
+  if (apl) {
+    ExpressionPrettyPrinter printer{out};
+    *out << ".*";
+  } else {
+    *out << "<null>";
```
Comment (Josipmrden) [inline src/query/frontend/ast/pretty_print.cpp:134]: why would it be null?

File: src/query/frontend/opencypher/grammar/Cypher.g4 [hunk 3]
```diff
@@ -290,6 +291,8 @@ patternComprehension : '[' ( variable '=' )? relationshipsPattern ( WHERE expres
 
 propertyLookup : '.' ( propertyKeyName ) ;
 
+allPropertiesLookup : '.' '*' ;
```
Comment (Josipmrden) [inline src/query/frontend/opencypher/grammar/Cypher.g4:294]: we have ASTERISK

File: src/query/interpret/eval.hpp [hunk 4]
```diff
@@ -466,7 +469,101 @@ class ExpressionEvaluator : public ExpressionVisitor<TypedValue> {
...
+        return TypedValue(result, ctx_->memory);
+      }
+      case TypedValue::Type::Duration: {
+        const auto &dur = expression_result.ValueDuration();
+        result.emplace("day", TypedValue(dur.Days(), ctx_->memory));
+        result.emplace("hour", TypedValue(dur.SubDaysAsHours(), ctx_->memory));
+        result.emplace("minute", TypedValue(dur.SubDaysAsMinutes(), ctx_->memory));
+        result.emplace("second", TypedValue(dur.SubDaysAsSeconds(), ctx_->memory));
+        result.emplace("millisecond", TypedValue(dur.SubDaysAsMilliseconds(), ctx_->memory));
+        result.emplace("microseconds", TypedValue(dur.SubDaysAsMicroseconds(), ctx_->memory));
```
Comment (Josipmrden) [inline src/query/interpret/eval.hpp:511]: is it a typo that for localtime you have millisecond and for duraation millisecond? or "s" is intentional

File: tests/benchmark/query/eval.cpp [hunk 5]
```diff
@@ -60,6 +60,8 @@ BENCHMARK_TEMPLATE(MapLiteral, NewDeleteResource)->Range(512, 1U << 15U)->Unit(b
 
 BENCHMARK_TEMPLATE(MapLiteral, MonotonicBufferResource)->Range(512, 1U << 15U)->Unit(benchmark::kMicrosecond);
 
+// TODO ante benchmark template for MapProjectionLiteral
```
Comment (Josipmrden) [inline tests/benchmark/query/eval.cpp:63]: what's this

**Review by Josipmrden — COMMENTED** [review]
small number of comments

File: src/query/interpret/eval.hpp (hunk 4 above)
Comment (antepusic) [reply src/query/interpret/eval.hpp:511]: Good catch! It should always be in the singular in line with the [temporal type property names](https://memgraph.com/docs/memgraph/reference-guide/data-types#temporal-types) and the rest of my code.

**Review by Josipmrden — APPROVED** [review]

**Review by Josipmrden — APPROVED** [review]

**Comment by antepusic** [comment]
@kgolubic I’ve opened PR [#888](https://github.com/memgraph/docs/pull/888) on the docs side concerning this PR

**Pull request merged**